pkg/logs/internal/util/adlistener/ad.go

```diff
@@ -0,0 +1,137 @@
+	// perform a non-blocking add to the channel
+	select {
+	case adMetaSchedulerCh <- sch:
+	default:
+	}
+}
+
+// ADListener implements pkg/autodiscovery/scheduler/Scheduler.
+//
+// It proxies Schedule and Unschedule calls to its parent, and also handles
+// delayed availability of the AD MetaScheduler.
+//
+// This must be a distinct type from schedulers, since both types implement
+// interfaces with different Stop methods.
+type ADListener struct {
+	// name is the name of this listener
+	name string
+
+	// schedule and unschedule are the functions to which Schedule and
+	// Unschedule calls should be proxied.
+	schedule, unschedule func([]integration.Config)
+
+	// adMetaScheduler is nil to begin with, and becomes non-nil after
+	// SetADMetaScheduler is called.
+	adMetaScheduler *scheduler.MetaScheduler
+
+	// cancelRegister cancels efforts to register with the AD MetaScheduler
+	cancelRegister context.CancelFunc
+
+	// registered is true if this listener has registered
+	registered bool
+
+	// registeredCond is a condition variable for 'registered'
+	registeredCond sync.Cond
+}
+
+var _ scheduler.Scheduler = &ADListener{}
+
+// NewADListener creates a new ADListener, proxying schedule and unschedule calls to
+// the given functions.
+func NewADListener(name string, schedule, unschedule func([]integration.Config)) *ADListener {
+	return &ADListener{
+		name:           name,
+		schedule:       schedule,
+		unschedule:     unschedule,
+		registeredCond: sync.Cond{L: &sync.Mutex{}},
+	}
+}
+
+// StartListener starts the ADListener.  It will subscribe to the MetaScheduler as soon
+// as it is available
+func (l *ADListener) StartListener() {
+	ctx, cancelRegister := context.WithCancel(context.Background())
+	go func() {
+		// wait for the scheduler to be set, and register once it is set
+		select {
+		case sch := <-adMetaSchedulerCh:
+			l.adMetaScheduler = sch
+			l.adMetaScheduler.Register(l.name, l)
+			// put the value back in the channel, in case it is needed again
+			SetADMetaScheduler(sch)
+
+			l.registeredCond.L.Lock()
+			l.registered = true
+			l.registeredCond.Broadcast()
+			l.registeredCond.L.Unlock()
+
+		case <-ctx.Done():
+		}
+	}()
+
+	l.cancelRegister = cancelRegister
+}
+
+// StopListener stops the ADListener
+func (l *ADListener) StopListener() {
+	l.cancelRegister()
+	if l.adMetaScheduler != nil {
+		l.adMetaScheduler.Deregister("logs")
+	}
+}
+
+// Stop implements pkg/autodiscovery/scheduler.Scheduler#Stop.
+func (l *ADListener) Stop() {}
+
+// Schedule implements pkg/autodiscovery/scheduler.Scheduler#Schedule.
+func (l *ADListener) Schedule(configs []integration.Config) {
+	l.schedule(configs)
+}
+
+// Unschedule implements pkg/autodiscovery/scheduler.Scheduler#Unschedule.
+func (l *ADListener) Unschedule(configs []integration.Config) {
+	l.unschedule(configs)
+}
+
+// waitForRegistration waits until this listener is registered
+func (l *ADListener) waitForRegistration() {
```

Comment:
is this function (and the lock/Cond) only used for UTs?

Reply:
It appears so.  If I recall, I wanted to use this _instead_ of refilling the channel, but context and cond vars don't play well together.  I'll remove the condvar.